### ISSUE TITLE
LBFGS does not work with SVI

### ISSUE THREAD

**Issue by fehiepsi** [uber/pyro#1519]
LBFGS requires us define a `closure` to be called in its `step` method. Defining a closure to call it with `loss_and_grads` is easy but I can't figure out how to incorporate it with the way PyroOptim handles parameters (creating an optimizer for each param).

**Comment by fehiepsi** [comment]
@eb8680 will we support it? LBFGS is a very convenient optimizer for models with small dataset! I have [used it to optimize a lot of models](https://github.com/fehiepsi/rethinking-pyro/blob/master/notebooks/rethinking.py#L49). Many of them can not be trained with SVI.

**Comment by eb8680** [comment]
@fehiepsi sorry, I only closed the issue because I thought it was resolved by the PR #1518. 

`pyro.optim.PyroOptim` is a klugy wrapper for handling varying amounts or orders of parameters.  I don't see how we could modify it to take a closure argument, but I also think the way you're using `torch.optim` in your linked example is fine - if anything, we should be encouraging users to use that idiom instead of `svi.step`, since it's more explicit.
